File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/OptionStatementSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class OptionStatementSyntax : DeclarationStatementSyntax
	{
		internal readonly KeywordSyntax _optionKeyword;

		internal readonly KeywordSyntax _nameKeyword;

		internal readonly KeywordSyntax _valueKeyword;

		internal static Func<ObjectReader, object> CreateInstance;

		internal KeywordSyntax OptionKeyword => _optionKeyword;

		internal KeywordSyntax NameKeyword => _nameKeyword;

		internal KeywordSyntax ValueKeyword => _valueKeyword;

		internal OptionStatementSyntax(SyntaxKind kind, KeywordSyntax optionKeyword, KeywordSyntax nameKeyword, KeywordSyntax valueKeyword)
			: base(kind)
		{
			base._slotCount = 3;
			AdjustFlagsAndWidth(optionKeyword);
			_optionKeyword = optionKeyword;
			AdjustFlagsAndWidth(nameKeyword);
			_nameKeyword = nameKeyword;
			if (valueKeyword != null)
			{
				AdjustFlagsAndWidth(valueKeyword);
				_valueKeyword = valueKeyword;
			}
		}

		internal OptionStatementSyntax(SyntaxKind kind, KeywordSyntax optionKeyword, KeywordSyntax nameKeyword, KeywordSyntax valueKeyword, ISyntaxFactoryContext context)
			: base(kind)
		{
			base._slotCount = 3;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(optionKeyword);
			_optionKeyword = optionKeyword;
			AdjustFlagsAndWidth(nameKeyword);
			_nameKeyword = nameKeyword;
			if (valueKeyword != null)
			{
				AdjustFlagsAndWidth(valueKeyword);
				_valueKeyword = valueKeyword;
			}
		}

		internal OptionStatementSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax optionKeyword, KeywordSyntax nameKeyword, KeywordSyntax valueKeyword)
			: base(kind, errors, annotations)
		{
			base._slotCount = 3;
			AdjustFlagsAndWidth(optionKeyword);
			_optionKeyword = optionKeyword;
			AdjustFlagsAndWidth(nameKeyword);
			_nameKeyword = nameKeyword;
			if (valueKeyword != null)
			{
				AdjustFlagsAndWidth(valueKeyword);
				_valueKeyword = valueKeyword;
			}
		}

		internal OptionStatementSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 3;
			KeywordSyntax keywordSyntax = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax != null)
			{
				AdjustFlagsAndWidth(keywordSyntax);
				_optionKeyword = keywordSyntax;
			}
			KeywordSyntax keywordSyntax2 = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax2 != null)
			{
				AdjustFlagsAndWidth(keywordSyntax2);
				_nameKeyword = keywordSyntax2;
			}
			KeywordSyntax keywordSyntax3 = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax3 != null)
			{
				AdjustFlagsAndWidth(keywordSyntax3);
				_valueKeyword = keywordSyntax3;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_optionKeyword);
			writer.WriteValue(_nameKeyword);
			writer.WriteValue(_valueKeyword);
		}

		static OptionStatementSyntax()
		{
			CreateInstance = (ObjectReader o) => new OptionStatementSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(OptionStatementSyntax), (ObjectReader r) => new OptionStatementSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.OptionStatementSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _optionKeyword, 
				1 => _nameKeyword, 
				2 => _valueKeyword, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new OptionStatementSyntax(base.Kind, newErrors, GetAnnotations(), _optionKeyword, _nameKeyword, _valueKeyword);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new OptionStatementSyntax(base.Kind, GetDiagnostics(), annotations, _optionKeyword, _nameKeyword, _valueKeyword);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitOptionStatement(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/OptionStatementSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class OptionStatementSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.DeclarationStatementSyntax
		Friend ReadOnly _optionKeyword As KeywordSyntax

		Friend ReadOnly _nameKeyword As KeywordSyntax

		Friend ReadOnly _valueKeyword As KeywordSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property NameKeyword As KeywordSyntax
			Get
				Return Me._nameKeyword
			End Get
		End Property

		Friend ReadOnly Property OptionKeyword As KeywordSyntax
			Get
				Return Me._optionKeyword
			End Get
		End Property

		Friend ReadOnly Property ValueKeyword As KeywordSyntax
			Get
				Return Me._valueKeyword
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal optionKeyword As KeywordSyntax, ByVal nameKeyword As KeywordSyntax, ByVal valueKeyword As KeywordSyntax)
			MyBase.New(kind)
			MyBase._slotCount = 3
			MyBase.AdjustFlagsAndWidth(optionKeyword)
			Me._optionKeyword = optionKeyword
			MyBase.AdjustFlagsAndWidth(nameKeyword)
			Me._nameKeyword = nameKeyword
			If (valueKeyword IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(valueKeyword)
				Me._valueKeyword = valueKeyword
			End If
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal optionKeyword As KeywordSyntax, ByVal nameKeyword As KeywordSyntax, ByVal valueKeyword As KeywordSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind)
			MyBase._slotCount = 3
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(optionKeyword)
			Me._optionKeyword = optionKeyword
			MyBase.AdjustFlagsAndWidth(nameKeyword)
			Me._nameKeyword = nameKeyword
			If (valueKeyword IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(valueKeyword)
				Me._valueKeyword = valueKeyword
			End If
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal optionKeyword As KeywordSyntax, ByVal nameKeyword As KeywordSyntax, ByVal valueKeyword As KeywordSyntax)
			MyBase.New(kind, errors, annotations)
			MyBase._slotCount = 3
			MyBase.AdjustFlagsAndWidth(optionKeyword)
			Me._optionKeyword = optionKeyword
			MyBase.AdjustFlagsAndWidth(nameKeyword)
			Me._nameKeyword = nameKeyword
			If (valueKeyword IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(valueKeyword)
				Me._valueKeyword = valueKeyword
			End If
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 3
			Dim keywordSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax)
				Me._optionKeyword = keywordSyntax
			End If
			Dim keywordSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax1 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax1)
				Me._nameKeyword = keywordSyntax1
			End If
			Dim keywordSyntax2 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax2 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax2)
				Me._valueKeyword = keywordSyntax2
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitOptionStatement(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.OptionStatementSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Select Case i
				Case 0
					greenNode = Me._optionKeyword
					Exit Select
				Case 1
					greenNode = Me._nameKeyword
					Exit Select
				Case 2
					greenNode = Me._valueKeyword
					Exit Select
				Case Else
					greenNode = Nothing
					Exit Select
			End Select
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._optionKeyword, Me._nameKeyword, Me._valueKeyword)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OptionStatementSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._optionKeyword, Me._nameKeyword, Me._valueKeyword)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._optionKeyword, IObjectWritable))
			writer.WriteValue(DirectCast(Me._nameKeyword, IObjectWritable))
			writer.WriteValue(DirectCast(Me._valueKeyword, IObjectWritable))
		End Sub
	End Class
End Namespace